### DISCUSSION TITLE
Require Developer Certificate of Origin to All Commits

### DISCUSSION THREAD

**Discussion by anshgoyalevil** [wagtail/wagtail#10322] [General]
In open source projects, a lot of contributors tend to add the lines of code from outer sources. This makes the project subject to legal rules.

The Developer Certificate of Origin (DCO) is a lightweight way for contributors to certify that they wrote or otherwise have the right to submit the code they are contributing to the project. Here is the full [text of the DCO](https://developercertificate.org/), reformatted for readability:

> By making a contribution to this project, I certify that:
>
> (a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or
>
> (b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or
>
> (c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.
>
> (d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

**Describe the solution you'd like**

Contributors sign-off that they adhere to these requirements by adding a Signed-off-by line to commit messages.
```
This is my commit message

Signed-off-by: Random J Developer <random@developer.example.org>
```
Git even has a -s command line option to append this automatically to your commit message:
```
$ git commit -s -m 'This is my commit message'
```

For this to be enforced among all commits, a DCO bot can be added.
Once installed, this integration will create a [check](https://developer.github.com/v3/checks/runs/) indicating whether or not commits in a Pull Request do not contain a valid Signed-off-by line.

**Approach to be followed (optional)**
- Go to https://probot.github.io/apps/dco/
- Click ```Add to Github```
- Select the org and repo where it has to be enabled.

**Additional context**
This feature is used by many open source projects since maintaining legal contracts is a good practice.
Example project which use this:
- https://github.com/meshery/meshery.io
- https://github.com/meshery/meshery

Example Check: https://github.com/meshery/meshery/pull/7259/checks?check_run_id=11720541887

**Comment by anshgoyalevil** [comment]
@thibaudcolas Please give your inputs over this.

**Comment by vsalvino** [comment]
Thanks for bringing this up. We did discuss it at the most recent Core Team meeting and came to the conclusion that it would be beneficial to put similar language into our contributing guide — but we do not plan to enforce DCO in commits or pull requests, following the precedent set by Python and Django.

**Comment by anshgoyalevil** [comment]
@vsalvino 

Alright. Should we close this discussion?

Btw. The whole Linux Open Source World uses it. Maybe, they enforce stricter rules.

**Comment by thibaudcolas** [comment]
Thanks for bringing this up! If we ever were to go for something like this, my preference would be for a contributor license agreement to be signed upfront rather than automated verification of every commit – which feels very heavyweight.

Ultimately for a CMS I don’t see it as very likely that we would receive submissions of code that could be valuable enough to be covered under patents (valid and enforceable ones), in a way that would make anyone come after "Wagtail". So the biggest risk would be someone submitting GPL code without realising.
